### ISSUE TITLE
Generating data by schema.

### ISSUE THREAD

**Issue by lk-geimfari** [lk-geimfari/mimesis#199]
We have implemented very primitive generator by schema:

```python
from mimesis.schema import Schema
from mimesis import constants as c

schema = Schema(locale=c.EN)

# We will use format 'provider.method'
# Param 'schema' should be dict or path to json file.
result = schema.load(schema={
    'username': 'personal.username',
    'password': 'personal.password',
    'full_name': 'datetime.full_name',
}).create()
```

But is not better solutions. @sobolevn suggest much better solutions using Lazy objects`, which looks like that:

```python
>>> from custom.utils import format_phone
>>> from mimesis.schema import Schema, fields
>>> schema = Schema('en')

>>> schema.load(schema={
...     "id": fields.cryptographic.uuid(version=4),
...     "name":  fields.personal.full_name(gender='female'),
...     "version": fields.development.version(semantic=True),
...     "phone": format_phone(fields.personal.phone),
... }).create(iterations=2)
```

I like second solution too.

**Comment by lk-geimfari** [comment]
I don't know how to be with arguments of functions. Maybe:
```python
schema={
    # equal to Personal(c.EN).username(gender='female')
    'username': 'personal.username.female',
}
```

**Comment by lk-geimfari** [comment]
What do you think? It's can be useful?

**Comment by lk-geimfari** [comment]
@Valerievich @sobolevn Look, please.

**Comment by lk-geimfari** [comment]
I have already implemented basic functionality. It remains a little to finish.

**Comment by sobolevn** [comment]
Looks good! But how would you pass params to `'personal.password'` for example?

**Comment by Valerievich** [comment]
One more pretty adorable way to generate data I think.

**Comment by lk-geimfari** [comment]
@sobolevn That is a real trouble. I would suggest `personal.surname.female`, where `.female` is first argument, but it's looks not so good if method has many arguments. We'll wait proposals.

**Comment by lk-geimfari** [comment]
Early version is already done: [mimesis/schema.py]( https://github.com/lk-geimfari/mimesis/blob/master/mimesis/schema.py)

**Comment by lk-geimfari** [comment]
How about following format?

```json
{
   "surname":{
      "provider": "personal",
      "args": {
         "gender": "female"
      }
   }
}
```

**Comment by sobolevn** [comment]
I think that arguments are really important. They should be both:

1. Easy to change
2. Easy to write

Consider this situation: I want to have user with different age https://github.com/lk-geimfari/mimesis/blob/master/mimesis/providers/personal.py#L29

**Comment by lk-geimfari** [comment]
I'm absolutely agree. But i cannot evaluate the complexity of implementation immediately. For example [this](https://github.com/lk-geimfari/mimesis/issues/199#issuecomment-319303881) looks very difficult for me and i have not idea how we can do it dynamically.

**Comment by lk-geimfari** [comment]
@Valerievich It's [done](https://github.com/lk-geimfari/mimesis#generate-data-by-schema). We need to implement support of arguments.

**Comment by Valerievich** [comment]
@lk-geimfari can u give me a couple examples of arguments please?

**Comment by sobolevn** [comment]
I don't like current implementation. It breaks one major rule: "everything is an object". Our fields right now are not objects in general case. They are strings. 

So it could break a lot of things for the end user. Imagine a user has some sort of logic to reformat phone numbers to his specific needs. Like: `reformat_phone(value)`. How is it possible with the current implementation? Or any other functions/classes/etc which wraps values.

What I suggest?

## Lazy objects (or generator)

In my opinion, we should create `LazyField` wrapper to wrap any other existing field. And a special `fields` container with all the existing fields wrapped into `LazyField`. So, how would it work?

>>> from custom.utils import format_phone
>>> from mimesis.schema import Schema, fields
>>> schema = Schema('en')

>>> schema.load(schema={
...     "id": fields.cryptographic.uuid(version=4),
...     "name":  fields.personal.full_name(gender='female'),
...     "version": fields.development.version(semantic=True),
...     "phone": format_phone(fields.personal.phone),
... }).create(iterations=2)

On each iteration lazy objects (or generator) generates new value. User has all the control, code is more pythonic. 

## Considerations

Do you have any ideas? Am I missing something?

**Comment by lk-geimfari** [comment]
@sobolevn Of course it's looks much better, than current implementation. I have only one question: How we can generate data by schema.json? Or, maybe is does not matter, cuz i have really like idea with `fields`.

**Comment by lk-geimfari** [comment]
@sobolevn  Can you explain, please, how to implement LazyField based data generator by schema? I mean steps. Maybe you have link to similar theme? I want to try implement it and close this issue on this week. Thanks you.

**Comment by sobolevn** [comment]
I came with even better idea: why now implementing custom provider for `factory_boy`? 
It already has all the stuff we need!

**Comment by lk-geimfari** [comment]
@sobolevn I never work with this library, but i'll try to figure out how to do it.

**Comment by lk-geimfari** [comment]
Unfortunately, I did not understand how to add the ability to pass arguments. @sobolevn  Can you look at this issue when you'll have free time, please?

**Comment by sobolevn** [comment]
Sure!

**Comment by samuarl** [comment]
The randomness distribution in schema generations seems off. Even from many iterations the resultant data usually only contains a handful of unique values per field.

```python
from mimesis.schema import Schema
import pandas as pd

schema = Schema('en')

data = schema.load(schema={
    "Name": "personal.name",
    "Surname": "personal.surname",
    "Username": "personal.username"
}).create(iterations=10000)

# Value counts and markdown tables
df = pd.DataFrame(data)

for name, series in df.iteritems():
    theader = f'{name}|Count'
    trow = ''
    for value, count in series.value_counts().iteritems():
        trow += f'{value}|{count}\n'
    print(f'{theader}\n-|-\n{trow}')
```
**10000 Iterations**

Name|Count
-|-
Wynell|6456
Nakita|1436
Spring|1218
Yuko|890

Surname|Count
-|-
Shannon|6435
Harrell|1447
Tyler|1212
Vincent|906

Username|Count
-|-
Vaultier_1956|6445
Compatriot_2055|1452
Dervish.1941|1205
edsel.1825|898

**Comment by lk-geimfari** [comment]
@samuarl It's strange, because we have disabled seed by default.

**Comment by lk-geimfari** [comment]
@samuarl I run this script on my laptop and everything is okay.

**Comment by lk-geimfari** [comment]
Implemented in 53c8741930fe0c79e605d53817e4ef4bcead0766.
